File: Sorting_Algo/Insertion.cpp
// time complexity = 0(n2) for worst and o(n) for best case


#include<iostream>
using namespace std;

void insertionSort(int arr[], int n){
    for(int i = 1; i < n; i++){
        int j = i;
        while(j > 0 && arr[j - 1] > arr[j]){
            int temp = arr[j - 1];
            arr[j-1] = arr[j];
            arr[j] = temp;
            j--;
        }
    }
}

void printArray(int arr[], int n){
    for(int i = 0; i < n; i++){
        cout << arr[i] << " ";
    }
    cout << endl;
}

int main(){
    int arr[] = {34,23,56,11,67};
    int n = sizeof(arr) / sizeof(arr[0]);
    cout << "Before sorting array is:" << endl;
    printArray(arr, n);

    insertionSort(arr, n);

    cout << "After insertion sort array is:" << endl;
    printArray(arr, n);
}

File: sortingjava/Insertion.java
// time complexity = 0(n2) for worst and o(n) for best case

package sortingjava;

public class Insertion {
     public static void insertionSort(int nums[]){
        int n = nums.length;
        for(int i = 1; i < n; i++){
           int curr = nums[i];
           int prev = i - 1;
           
           while(prev >= 0 && nums[prev] > curr){
            nums[prev + 1] = nums[prev];
            prev--;
           } 

           nums[prev + 1] = curr;

        }
    }
    

    public static void printArray(int nums[]){
        for(int i = 0; i < nums.length; i++){
            System.out.print(nums[i] + " ");
        }
        System.out.println();
    }
    public static void main(String[] args) {
        int[] nums = {3,2,8,5,6};

        System.out.println("Array before sorting:");
        printArray(nums);

        System.out.println("Array after applying insertion sort");
        insertionSort(nums);
        printArray(nums);
        
    }
    
}
